tests/test_str_t.c

```diff
@@ -47,14 +42,83 @@ void test_creating_new_raw_parts() {
   str_free(str);
 }
 
+void test_copy_to_zero_terminated() {
+  str_t* str = str_new_from_zero_terminated("Hi!");
+  char dest[4] = {'o', 'o', 'o', 'o'};
+
+  str_err_t res = str_copy_to_zero_terminated(str, dest, sizeof(dest));
+
+  TEST_ASSERT_EQUAL_INT(STR_T_OK, res);
+  TEST_ASSERT_EQUAL_STRING("Hi!", dest);
+
+  str_free(str);
+}
+
+void test_copy_to_zero_terminated_error() {
+  str_t* str = str_new_from_zero_terminated("Hi!");
+  char dest[3];
+
+  str_err_t res = str_copy_to_zero_terminated(str, dest, sizeof(dest));
+  TEST_ASSERT_EQUAL_INT(STR_T_BUFFER_TOO_SMALL, res);
+
+
+  str_free(str);
+}
+
+
+void test_str_copy() {
+  str_t* source = str_new_from_zero_terminated("Hi!");
+  str_t* copy = str_copy(source);
+
+  TEST_ASSERT_EQUAL_INT(3, copy->len);
+  TEST_ASSERT_EQUAL_STRING_LEN("Hi!", copy->ptr, 3);
+
+  TEST_ASSERT_NOT_EQUAL(source, copy);
+
+  str_free(source);
+  str_free(copy);
+}
+
+void test_str_concat() {
+  str_t* hello = str_new_from_zero_terminated("Hello ");
+  str_t* matey = str_new_from_zero_terminated("matey!");
+
+  str_t* hello_matey = str_concat(hello, matey);
+
+  TEST_ASSERT_EQUAL_INT(12, hello_matey->len);
+  TEST_ASSERT_EQUAL_STRING_LEN("Hello matey!", hello_matey->ptr, 12);
+
+  str_free(hello);
+  str_free(matey);
+  str_free(hello_matey);
+}
+
+void test_str_move() {
+  // Guessing at the semantics. This seems dangerous as we could
+  // accidentally free an array twice;
+
+  str_t* source = str_new_from_zero_terminated("hi");
+  str_t dest;
```

Comment:
So far we've been allocating `str_t` types on the heap. But this one is on the stack and after calling `str_move` it carries a pointer to heap memory for the string.

Reply:
True, but it's fine the way it's being used for the test. 

I'm struggling to guess what the purpose of move could be, tbh. I figure that (as usual) it's up to the client to understand what happens to the memory.